### DISCUSSION TITLE
Bringing Back the Layout Switcher Feature

### DISCUSSION THREAD

**Discussion by taufik-nurrohman** [mecha-cms/mecha#205] [Ideas]
### Why?

[![why-coki-muslim](https://user-images.githubusercontent.com/1669261/160343646-d6a6dd72-a03e-4bd8-acc0-5236f5e7f172.gif)](https://youtu.be/smC_GdwjdWY?t=635)

After [2 years, 2 months and 28 days have passed since the release of Mecha version 2.2.0](https://mecha-cms.com/article/version-2-2-0), I am having intolerable difficulties regarding synchronization between the local layout file structure and the remote repository associated with that layout. It&rsquo;s very difficult for me to develop automatic update features on the layout because when the layout has been uploaded, its folder name will disappear and all files in it will merge into one in the `.\lot\layout` folder. Unlike extensions which will be uploaded including the folder, so that the mapping between extension name and remote repository associated with that extension can be retained.

#### Typical Extension File Structure Per v2.6.4

~~~ txt
.\
└── lot\
    └── x\
        └── foo-bar\ ← this is `foo-bar` extension, obviously!
            ├── engine\
            ├── lot\
            ├── about.page
            ├── index.php
            └── state.php
~~~

#### Typical Extension Repository Structure Per v2.6.4

~~~ txt
https://github.com/
└── mecha-cms/
    └── x.foo-bar/
        └── foo-bar/
            ├── engine/
            ├── lot/
            ├── about.page
            ├── index.php
            └── state.php
~~~

#### Typical Layout File Structure Per v2.6.4

~~~ txt
.\
└── lot\
    └── layout\ ← what layout is this?
        ├── asset\
        ├── about.page
        ├── index.php
        └── state.php
~~~

#### Typical Layout Repository Structure Per v2.6.4

~~~ txt
https://github.com/
└── mecha-cms/
    └── layout.foo-bar/
        ├── asset/
        ├── about.page
        ├── index.php
        └── state.php
~~~

At first, I tried to make [a `git` property proposal in `about.page` file](https://github.com/mecha-cms/x.panel/blob/v2.8.1/panel/engine/r/lot/pages/layout.php#L92) to solve this problem. Hopefully, this can also be applied to third party extensions uploaded outside of <https://github.com/mecha-cms> organization, so that they can tell [panel](https://github.com/mecha-cms/x.panel) about their repository name in an `about.page` file&rsquo;s `git` property to be fetched.

~~~ yaml
---
title: Layout Title
description: Layout description.
git: 'taufik-nurrohman/layout.foo-bar'
type: Markdown
...

Lorem ipsum dolor sit amet.
~~~

This should works fine if the `git` property exists, but gets a bit hacky when I have to guess the repository name by creating a folder name via the layout title. Layout titles should be freely defined and unrelated to the internal structure of the file system. Repository name errors will occur when an uninformed developer decides to change his layout title without changing his remote repository name.

### Using the `y.` Prefix

Instead of using the `layout.` prefix, I&rsquo;d probably use the `y.` prefix in the future to accompany the `.\lot\x` folder, so we&rsquo;ll have unique folders named `.\lot\x` and `.\lot\y`, just like in algebraic mathematics. Here, _x_ will become an abbreviation for _extension_ (noun) and _extend_ (verb), and the `.\lot\x` folder will be used to store extensions. And _y_ will become an abbreviation for _layout_ (noun) and _yield_ (verb), and the `.\lot\y` folder will be used to store layouts.

~~~ txt
.\
└── lot\
    ├── asset\
    ├── page\
    ├── x\
    │   ├── extension-1\
    │   ├── extension-2\
    │   ├── extension-3\
    │   └── …
    └── y\
        ├── layout-1\
        ├── layout-2\
        ├── layout-3\
        └── …
~~~

### The Rise of `X` and `Y` Class

~~I plan to remove the `layout` extension from the core and make it as a built-in non-removable feature instead.~~ 👎 This feature should be very simple and easy to develop.

Right now, the only idea that comes to my mind is simply to rename `Layout` class to `Y`, and to add class `X` as a companion to `Y` which is currently useless.

Below is an example usage of the `Y` class. Pretty much the same as when using `Layout` class.

~~~ php
!Y::path('alert') && Y::set('alert', $path_to_alert_layout);
~~~

~~~ php
<form method="get">
  <?= Y::alert(); ?>
  <p>
    <input name="query" type="text"> <button type="submit">
      <?= i('Search'); ?>
    </button>
  </p>
</form>
~~~

The only method that can be useful for `X` class is probably just `has`.

~~~ php
if (X::has('page', true)) {
    // Page extension does exist and is active!
}
~~~

### Advantages?

 - Makes it possible to have a layout repository containing multiple layout folders at once, great for creating collections or reviving a list of themes from certain dead web platforms.
 - No need to guess repository name from the layout title.
 - Opens up opportunities for new ideas about loading different layout on specific pages within one website.
 - Opens up opportunities for new ideas regarding &ldquo;child theme&rdquo; features like in [WordPress](https://developer.wordpress.org/themes/advanced-topics/child-themes).

### Disadvantages?

 - Adds complexity.
 - Forced me to have a useless class named `X`.
 - Makes the layout asset path longer.

---

Another ~~why~~ way to make `X` and `Y` classes more viable to exist is to extend them to the `State` class. As there&rsquo;s a `state.php` file ​​in the core, and also a `state.php` file for each extension and layout folder makes more sense to me.

~~~ php
class X extends State {} 👎
class Y extends State {} 👎
~~~

_To be continued._

**Comment by taufik-nurrohman** [comment]
To keep class `X` from getting in, it&rsquo;s probably best not to provide `X` and `Y` classes at all :sunglasses:

**Comment by taufik-nurrohman** [comment]
Having automatic custom layout feature is probably a bad idea. It would be better if custom layouts were available as custom fields. But it will probably be available in a separate extension:

~~~ yaml
---
title: Page Title
description: Page description goes here.
layout: page/index
type: Markdown
...

Lorem ipsum dolor sit amet.
~~~

Example extension to enable this feature:

~~~ php
Hook::set('route', function($content) {
    if (is_array($content) && isset($content[0]) && is_string($content[0])) {
        $page = $GLOBALS['page'] ?? new Page;
        if ($page->exist && $layout = $page->layout) {
            $content[0] = $layout;
        }
    }
    return $content;
}, 900);
~~~

**Reply by nurbaitifitriyani** [reply]
Saya setuju kalau ada sebuah ekstensi untuk opsi `layout`. Misalnya ada pengguna yang membutuhkan format postingan tertentu, seperti jurnal ilmiah yang harus punya bagian-bagian seperti abstrak, kata kunci, editor, ralat, dll atau postingan semacam toko online atau postingan semacam galeri foto/lukisan, tentu `layout` jadi berbeda-beda, terutama untuk yang mengejar `schema` (web page, blog posting, article, news article, journal, dll). Lagi, ini juga akan membuat opsi `panel` memberikan pilihan postingan tipe apa sehingga formulir isiannya juga menyesuaikan. Kalau hanya ada satu `layout`, tetapi memaksakan untuk bisa posting macam-macam, dengan kondisional PHP jadi `page.php` terlalu panjang juga.

*Sorry* ya Mas saya belum mampu untuk berkontribusi dalam donasi ke Mecha dalam waktu dekat ini. Saya baru saja selesai menamatkan kuliah. Nanti *kalau nasib saya sudah jelas*, *in sya-a Allah* saya akan merealisasikan niat saya tersebut. Saya memakai Mecha versi stabil rilis terbaru dan saya gunakan di situs pribadi saya, `selarakta.xyz`. Saya menyematkan tautan `mecha-cms.com` di salah satu laman saya di `selarakta.xyz/kolofon`. Terima kasih untuk segala bantuan dan diskusinya, saya tetap akan kerap datang di diskusi-diskusi `https://github.com/mecha-cms/mecha/discussions` karena pasti saya akan menemukan masalah dan ketidaktahuan saya atas masalah tersebut.

**Reply by taufik-nurrohman** [reply]
Ndak apa-apa _mba_, menjadi pengguna saja sudah sangat cukup untuk berkontribusi saat ini. Karena semakin banyak pemakai juga akan semakin meningkatkan nama Mecha di mesin pencari.

Kebetulan Saya cukup perfeksionis soal tata bahasa Indonesia. Saya termasuk tipe yang tidak nyaman mencampur-adukkan antara bahasa Indonesia dengan bahasa asing. Mungkin ke depannya Saya akan perlu belajar literasi bahasa juga dari _mba_.

**Reply by taufik-nurrohman** [reply]
This custom field is now built-in.

https://github.com/mecha-cms/x.layout/commit/6fcabcb8515532a50a8c36feb938fd658dbcba8f

**Comment by taufik-nurrohman** [comment]
The error page layout should not exist as a separate file. This layout can be expressed only in `page.php` or `pages.php` files to reduce the standard layout file structure specification.

~~~ php
<?= self::enter(); ?>
<?php if ($page->exist): ?>
  <article id="page:<?= $page->id; ?>">
    <h2>
      <?= $page->title; ?>
    </h2>
    <?= $page->content; ?>
  </article>
<?php else: ?>
  <article>
    <h2>
      <?= i('Error'); ?>
    </h2>
    <p>
      <?= i('Page does not exist.'); ?>
    </p>
  </article>
<?php endif; ?>
<?= self::exit(); ?>
~~~

~~~ php
<?= self::enter(); ?>
<?php if ($page->exist): ?>
  <?php if ($pages->count > 0): ?>
    <?php foreach ($pages as $page): ?>
      <article id="page:<?= $page->id; ?>">
        <h3>
          <a href="<?= $page->link ?? $page->url; ?>">
            <?= $page->title; ?>
          </a>
        </h3>
        <p>
          <?= $page->description; ?>
        </p>
      </article>
    <?php endforeach; ?>
  <?php else: ?>
    <?php if ($site->has('part')): ?>
      <p>
        <?= i('No more pages to show.'); ?>
      </p>
    <?php else: ?>
      <p>
        <?= i('No pages yet.'); ?>
      </p>
    <?php endif; ?>
  <?php endif; ?>
<?php else: ?>
  <article>
    <h2>
      <?= i('Error'); ?>
    </h2>
    <p>
      <?= i('Page does not exist.'); ?>
    </p>
  </article>
<?php endif; ?>
<?= self::exit(); ?>
~~~
